### PR TITLE
python examples using clustering 

### PR DESCRIPTION
**Pull request by damadux**
Adds 2 python scripts using clustering libraries from SciPy and Sklearn. The analysis clusters system calls based on their duration, then outputs a xy graph with matplotlib (x for system call duration, y for system call name).

### CODE REVIEW

File: examples/python/kMeanClusteringSciPy.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
+# Imports
```
Comment (tahini) [inline examples/python/kMeanClusteringSciPy.py:5]: Please add a copyright header at the top of the file. Also more comments for the users of this file, as some may go straight to this and not necessarily pass by the README beforehand. The header should explain what this file does, what trace it runs on and any setup required to run it properly.

File: README.md [hunk 2]
```diff
@@ -21,6 +21,7 @@ In the [utils](utils) directory are utility scripts that can be used as is with
...
  - **examples/javascript/[scriptedDataProvider.js](examples/javascript/scriptedDataProvider.js)** and **examples/python/[scriptedDataProvider.py](examples/python/scriptedDataProvider.py)**: Works with the [mpi_ring traceset](traces/mpi_ring.tgz), shows how to script one's own data provider to display in a time graph view, with arrows.
  - **examples/javascript/[statisticsDensityXY.js](examples/javascript/statisticsDensityXY.js)**: Works with any trace, draws a XY chart of the density of a particular event using the Statistics module.
+ - **examples/python/[kMeanClusteringSciPy.py](examples/python/kMeanClusteringSciPy.py)** and **examples/python/[kMeanClusteringSciPy.py](examples/python/kMeanClusteringSklearn.py)**: Works with kernel trace, uses the matplotlib and numpy libraries. **[kMeanClusteringSciPy.py](examples/python/kMeanClusteringSciPy.py)** also uses the scipy library, and **[kMeanClusteringSciPy.py](examples/python/kMeanClusteringSklearn.py)** also uses the scikit-learn library. In the run arguments, leave one number for the desired number of clusters for the kmean algorithm.
```
Comment (tahini) [inline README.md]: "Additionally <scipy> uses the scipy library, while <sklean> uses the scikit-learn library".

To describe the argument, look at the https://github.com/tahini/tracecompass-ease-scripting/blob/master/utils/javascript/exportUstSymbols.js script, where the header has the information for the expected arguments. It isn't clear here what it means. And give an example of usage, or a suggested value.

File: examples/python/kMeanClusteringSciPy.py (hunk 1 above)
Comment (tahini) [reply examples/python/kMeanClusteringSciPy.py:5]: You can add an MIT license like this:

```
# The MIT License (MIT)
#
# Copyright (C) 2019 - You
#
```

File: examples/python/kMeanClusteringSciPy.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+	# No duration is < than 0
+	pyplot.xlim(xmin=0)
+	pyplot.ylim(ymin=-2)
+
+syscallNameDistinct, syscallNameCount, syscallNameList, durations = extractAspects()
+codebook, distortion = runClustering(durations)
+syscallNameIndex, syscallName, durationData = prepareScatterPlot(syscallNameCount, codebook, syscallNameList)
+showPlot(syscallNameDistinct, syscallNameCount, durations, syscallNameList)
+printStatistics()
+pyplot.show()
```
Comment (tahini) [inline examples/python/kMeanClusteringSciPy.py:135]: With an argument of 3, I get the following error:
org.eclipse.ease.ScriptExecutionException: Traceback (most recent call last):
  File "workspace://test traces/kMeanClustringSciPy.py", line 130, in <module>
    if converter.can_convert(value):
  File "/usr/lib/python3.8/site-packages/matplotlib/pyplot.py", line 269, in show
    return _show(*args, **kw)
  File "/usr/lib/python3.8/site-packages/matplotlib/cbook/deprecation.py", line 413, in wrapper
    return func(*args, **kwargs)
  File "/usr/lib/python3.8/site-packages/matplotlib/backend_bases.py", line 3302, in show
    cls.mainloop()
  File "/usr/lib/python3.8/site-packages/matplotlib/backends/backend_qt5.py", line 1094, in mainloop
    signal.signal(signal.SIGINT, signal.SIG_DFL)
  File "/usr/lib/python3.8/signal.py", line 47, in signal
    handler = _signal.signal(_enum_to_int(signalnum), _enum_to_int(handler))
ValueError: signal only works in main thread

What am I doing wrong?

File: examples/python/kMeanClusteringSciPy.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+def showPlot(syscallNameDistinct, syscallNameCount, durations, syscallNameList):
+	fig, ax = pyplot.subplots()
+	formatter = FuncFormatter(formatDuration)
+	ax.xaxis.set_major_formatter(formatter)
+	# Titles
+	pyplot.title("System Call Clustering by Duration", fontsize=20,fontweight='bold')
+	pyplot.xlabel("System Call Duration (ns)", fontsize=14)
+	pyplot.ylabel("System Call Name", fontsize=14)
+
+	for i in range(int(argv[0])):
```
Comment (tahini) [inline examples/python/kMeanClusteringSciPy.py:123]: A good practice would be to start the script by verifying that an argument is present and print an error message if there is none, explaining the argument required.

**Review by tahini — CHANGES_REQUESTED** [review]
Thanks for the scripts! It's great to have example of ML libraries with trace data!

Some comments though.

Also, I know these are meant to be examples scripts, but I am not very familiar with ML and the output of those example scripts has no meaning to me:

For example, "{'ioctl': 0, 'splice': 1, 'sync_file_range2': 2, 'arm_fadvise64_64': 3, 'clock_gettime': 4, 'read': 5...", I don't know what it means. But I get errors, maybe there is supposed to be more outputs after that? Anyway, if you could provide some insight in the output, something that will help lay[wo]men make sense of this example, that would be great!

**Comment by tahini** [comment]
With new updates with py4j vs jython scripting engine, these scripts are good examples of mixed engine scripts, and I managed to have them working properly with python2. So I'll re-work then with that in mind and merge this PR.

Then we can document that it works only with python2? or you tell me why it does not work with python 3. And let's fix it if possible!

**Comment by tahini** [comment]
@damadux Could you take a look at the damadux-master branch that I update in this repo? If it's ok with you, I'll merge it with master and close this PR. Thanks!

**Comment by damadux** [comment]
Looks good! I will test these with py4j and jython.

**Comment by tahini** [comment]
now merged. Thanks! Test it please, I'll let you figure out how to do so, from the doc and eventual output messages. If it's not clear, patches welcome!

**Comment by damadux** [comment]
Did you still have an error running these examples with Python 3? I had no issues with my machine.

**Comment by tahini** [comment]
Yes, but setting this: matplotlib.use("GTK3Agg")    fixed it. It was using Qt5 backend I think and there there is a signal that needs to be run from the UI thread. So it's not really python3 problem after all, but pyplot backend. The script in master has that fix.

**Pull request merged**